third-culture-software/bhima
# Feat: merge ODK central work

This PR adds in [ODK Central](https://docs.getodk.org/central-install/) functionality to the BHIMA server.  Here is what the ODK administration interface looks like:

![image](https://user-images.githubusercontent.com/896472/150951277-34cf068b-9eca-4363-a485-ad135ad18df1.png)

You'll need to set your own username/password and ODK Central URL to test it out.

This installation is very bare-bones.  The way this works is somewhat documented in https://github.com/IMA-WorldHealth/bhima/issues/6235#issuecomment-1014303259.

This PR is plucked from work done by @jniles and @mbayopanda in `sprint-prosani`.

Closes #6235.
Closes #6302.
Closes #6322.

---

**TESTING**  (based on session with @jniles )
- Use bhima test (rebuild db, etc)
- Go to Administration > ODK Settings
  - Enter the ODK Central login credentials in the left-most panel (ODK Integration Settings)
  - In the middle panel, click on
    - Synchronize enterprise settings
    - Synchronize users
    - Synchronize forms
- Go to Adminstration > User Management
   - In the action menu for Superuser, click on the [ODK QR Code] link to display the QR code (leave it up on your screen)
- On your mobile phone
  - If you have not installed ODK Collect, do that
  - In ODK Collect, click on the [Configure with QR code] button on the entry screen. This should bring up the "Test Enterprise" page
  - Click on [Get Blank Form].  It should show that it has downloaded a form
  - On the bottom right of the page, click on the [Get Selected] button.  This downloads the form and sends you back to the "Test Enterprise" page.
  - Click on [Fill Blank Form] and select the form.  This should start filling out the form.
  - Fill out the form
     - For Nombre des articles, enter 1
     - For the barcode, first go to Bhima (Stock Lots) and display the barcode for lot VITAMIN-A.  Then scan that barcode with your phone.  When you click next, it should show the lot you selected.
     - Finish the form ([X] Mark form as finalize, press [Save Form and Exit])
  - Back on the "Test Enterprise" page, click [Send Finalized Form].  Select the form you just completed and press [Send Selected] on the bottom right of the page.
  - Back on Bhima go to Administration > ODK Settings
  - Click on:  Synchronize Submissions [Sync]
  - Go to Stock > Stock Movements and verify that you have a new stock entry in Depot Tertiaire
